src/docfx/build/Build.cs

```diff
@@ -44,40 +43,79 @@ private static List<Document> GlobFiles(Context context, Docset docset)
                            .ToList();
         }
 
-        private static async Task<(List<Document> builtDocs, DependencyMap sourceDependencies)> BuildFiles(Context context, List<Document> files, TableOfContentsMap tocMap)
+        private static async Task<(List<Document> publishedFiles, DependencyMap sourceDependencies)> BuildFiles(Context context, List<Document> files, TableOfContentsMap tocMap)
         {
-            var builtDocs = new ConcurrentDictionary<Document, byte>();
-            var references = new ConcurrentDictionary<Document, byte>();
             var sourceDependencies = new ConcurrentDictionary<Document, List<DependencyItem>>();
             var buildScope = new HashSet<Document>(files);
 
-            await ParallelUtility.ForEach(
-                files,
-                async (file, buildChild) =>
+            var publishConflicts = new ConcurrentDictionary<string, ConcurrentBag<Document>>();
+            var publishedFiles = new ConcurrentDictionary<string, Document>();
+
+            await ParallelUtility.ForEach(files, BuildTheFile, ShouldBuildTheFile);
+
+            HandlePublishConflicts();
+
+            return (
+                publishedFiles.Values.OrderBy(d => d.OutputPath).ToList(),
+                new DependencyMap(sourceDependencies.OrderBy(d => d.Key.FilePath).ToDictionary(k => k.Key, v => v.Value)));
+
+            async Task BuildTheFile(Document file, Action<Document> buildChild)
+            {
+                var dependencyMap = await BuildFile(context, file, tocMap, buildChild);
+
+                foreach (var (souce, dependencies) in dependencyMap)
+                {
+                    sourceDependencies.TryAdd(souce, dependencies);
+                }
+            }
+
+            bool ShouldBuildTheFile(Document file)
+            {
+                if (!ShouldBuildFile(context, file, tocMap, buildScope))
+                {
+                    return false;
+                }
+
+                // Find publish URL conflicts
+                if (!publishedFiles.TryAdd(file.SiteUrl, file))
```

Comment:
>publishedFiles [](start = 21, length = 14)

remember to filter out the dependency repos' files

Reply:
+1

---
In reply to: [194342711](https://github.com/dotnet/docfx/pull/2868#discussion_r194342711) [](ancestors = 194342711)

Reply:
It will be covered in other PRs, when we have dependency tests

---
In reply to: [194599901](https://github.com/dotnet/docfx/pull/2868#discussion_r194599901) [](ancestors = 194599901,194342711)

Reply:
sure, how about add a TODO

---
In reply to: [194605099](https://github.com/dotnet/docfx/pull/2868#discussion_r194605099) [](ancestors = 194605099,194599901,194342711)